File: src/main/java/com/pschlup/ta/timeseries/TimeFrame.java
package com.pschlup.ta.timeseries;

import java.time.Duration;

@SuppressWarnings("unused")
public enum TimeFrame {

  M5(Duration.ofMinutes(5)),
  M10(Duration.ofMinutes(10)),
  M15(Duration.ofMinutes(15)),
  M30(Duration.ofMinutes(30)),
  H1(Duration.ofHours(1)),
  H2(Duration.ofHours(2)),
  H4(Duration.ofHours(4)),
  D(Duration.ofDays(1));

  private final Duration duration;

  TimeFrame(Duration duration) {
    this.duration = duration;
  }

  public Duration duration() {
    return this.duration;
  }

  public static TimeFrame of(Duration duration) {
    switch ((int)duration.toMinutes()) {
      case 5:
        return M5;
      case 10:
        return M10;
      case 15:
        return M15;
      case 30:
        return M30;
      case 60:
        return H1;
      case 120:
        return H2;
      case 240:
        return H4;
      case 1440:
        return D;
      default:
        throw new IllegalArgumentException(
            String.format("Unknown timeframe of %d minutes", duration.toMinutes()));
    }
  }
}

File: src/main/kotlin/com/pschlup/ta/timeseries/TimeFrame.kt
package com.pschlup.ta.timeseries

import java.time.Duration

enum class TimeFrame(val duration: Duration) {
  M5(Duration.ofMinutes(5)),
  M10(Duration.ofMinutes(10)),
  M15(Duration.ofMinutes(15)),
  M30(Duration.ofMinutes(30)),
  H1(Duration.ofHours(1)),
  H2(Duration.ofHours(2)),
  H4(Duration.ofHours(4)),
  D(Duration.ofDays(1));

  companion object {
    @JvmStatic
    fun of(duration: Duration): TimeFrame = when (duration.toMinutes().toInt()) {
      5 -> M5
      10 -> M10
      15 -> M15
      30 -> M30
      60 -> H1
      120 -> H2
      240 -> H4
      1440 -> D
      else -> throw IllegalArgumentException("Unknown timeframe of ${duration.toMinutes()} minutes")
    }
  }
}